idx/memory/tag_query.go

```diff
@@ -2,504 +2,240 @@ package memory
 
-// getInitialByMatch generates the initial resultset by executing the given match expression
-func (q *TagQueryContext) getInitialByMatch(expr kvRe, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
+	q.filters = make([]tagquery.MetricDefinitionFilter, len(q.query.Expressions)-1)
+	q.defaultDecisions = make([]tagquery.FilterDecision, len(q.query.Expressions)-1)
 
-	// shortcut if Regex == nil.
-	// this will simply match any value, like ^.+. since we know that every value
-	// in the index must not be empty, we can skip the matching.
-	if expr.Regex == nil {
-	VALUES1:
-		for _, ids := range q.index[expr.Key] {
-			for id := range ids {
-				select {
-				case <-stopCh:
-					break VALUES1
-				case idCh <- id:
-				}
-			}
+	i := 0
+	for _, cost := range costs {
+		if q.startWith < 0 && q.query.Expressions[cost.expressionIdx].RequiresNonEmptyValue() {
+			q.startWith = cost.expressionIdx
+		} else {
+			q.filters[i] = q.query.Expressions[cost.expressionIdx].GetMetricDefinitionFilter(idx.idHasTag)
+			q.defaultDecisions[i] = q.query.Expressions[cost.expressionIdx].GetDefaultDecision()
+			i++
 		}
-		close(idCh)
-		return
 	}
+}
 
-VALUES2:
-	for v, ids := range q.index[expr.Key] {
-		if !expr.Regex.MatchString(v) {
-			continue
-		}
+// getInitialIds asynchronously collects all ID's of the initial result set.  It returns:
+// a channel through which the IDs of the initial result set will be sent
+// a stop channel, which when closed, will cause it to abort the background worker.
+func (q *TagQueryContext) getInitialIds() (chan schema.MKey, chan struct{}) {
+	idCh := make(chan schema.MKey, 1000)
+	stopCh := make(chan struct{})
 
-		for id := range ids {
-			select {
-			case <-stopCh:
-				break VALUES2
-			case idCh <- id:
-			}
-		}
+	if q.query.Expressions[q.startWith].OperatesOnTag() {
+		q.getInitialByTag(idCh, stopCh)
+	} else {
+		q.getInitialByTagValue(idCh, stopCh)
 	}
 
-	close(idCh)
+	return idCh, stopCh
 }
 
-// getInitialByTagPrefix generates the initial resultset by creating a list of
-// metric IDs of which at least one tag starts with the defined prefix
-func (q *TagQueryContext) getInitialByTagPrefix(idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
+// getInitialByTagValue generates an initial ID set which is later filtered down
+// it only handles those expressions which involve matching a tag value:
+// f.e. key=value but not key!=
+func (q *TagQueryContext) getInitialByTagValue(idCh chan schema.MKey, stopCh chan struct{}) {
+	expr := q.query.Expressions[q.startWith]
 
-TAGS:
-	for tag, values := range q.index {
-		if !strings.HasPrefix(tag, q.tagPrefix) {
-			continue
-		}
+	q.wg.Add(1)
+	go func() {
+		defer close(idCh)
+		defer q.wg.Done()
+
+		keyPtr, found := expr.GetKeyPtr()
+		if !found {
+			// the key value has never been interned by the index, so it is safe to
+			// assume it does not exist. this means this query must result in an
+			// empty result set
```

Comment:
why is the outcome so different from above where we said "so this expression will be cheap to evaluate"

Reply:
When building the initial result set we try to look up the metric definitions that match the expression which has chosen to be the one we start with. If we see that the key of this expression does not even exist in the index (it can't because it has not been interned) then we don't need to look anything up, because we can conclude directly that the result set will be empty.
The filter expressions above which are "cheap to evaluate" act according to the same logic. If a filter check whether some `metricX` has `tag=value`, but it already knows the value `tag` does not exist in the index at all because it's not been interned, then it doesn't even need to look at any of the tags or the tag index to conclude that `metricX` does not have `tag=value`.